server/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -203,6 +203,11 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
             // TODO this should be caught by lucene - EOF is almost certainly an index corruption
             throw new CorruptIndexException("Read past EOF while reading segment infos", "commit(" + commit + ")", eof);
         } catch (IOException exception) {
+            final String message = exception.getMessage();
+            // DataInput throws a generic IOException on odd data cases
+            if (message != null && (message.startsWith("Invalid vLong detected") || message.startsWith("Invalid vInt detected"))) {
```

Comment:
If we are going to do this, we should add a test that asserts these exception messages continue to be generated by Lucene (i.e., that we are not prone to them silently changing underneath us).

Reply:
@jasontedor I've addressed that in 2a18280